crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -347,18 +341,14 @@ impl AppService {
     /// # Returns
     /// This function may return a UIAA response, which should be checked for
     /// with [`Error::as_uiaa_response()`].
-    pub async fn register_user<'a>(
-        &self,
-        localpart: &'a str,
-        device_id: Option<&'a DeviceId>,
-    ) -> Result<()> {
+    pub async fn register_user(&self, localpart: &str, device_id: Option<&DeviceId>) -> Result<()> {
         if self.is_user_registered(localpart).await? {
             return Ok(());
         }
         let request = assign!(register::v3::Request::new(), {
-            username: Some(localpart),
-            login_type: Some(&register::LoginType::ApplicationService),
-            device_id,
+            username: Some(localpart.to_owned()),
+            login_type: Some(register::LoginType::ApplicationService),
+            device_id: device_id.map(ToOwned::to_owned),
```

Comment:
may I suggest we change our API then, too? The str is pretty awkward to work with if you have e.g. have an Option<String> coming in and we have to to_owned() them now anyways ...